maven-plugin/plugin-core/src/main/java/org/jvnet/jaxb/maven/resolver/tools/ReResolvingEntityResolverWrapper.java

```diff
@@ -12,33 +18,51 @@ public class ReResolvingEntityResolverWrapper implements EntityResolver {
 
 	public ReResolvingEntityResolverWrapper(EntityResolver entityResolver) {
 		if (entityResolver == null) {
-			throw new IllegalArgumentException(
-					"Provided entity resolver must not be null.");
+			throw new IllegalArgumentException("Provided entity resolver must not be null.");
 		}
 		this.entityResolver = entityResolver;
 	}
 
 	@Override
 	public InputSource resolveEntity(String publicId, String systemId)
 			throws SAXException, IOException {
-		// System.out.println(MessageFormat.format("Resolving publicId [{0}], systemId [{1}].",
-		// publicId, systemId));
-		final InputSource resolvedInputSource = this.entityResolver
-				.resolveEntity(publicId, systemId);
+		// System.out.println(MessageFormat.format("Resolving publicId [{0}], systemId [{1}].", publicId, systemId));
+		final InputSource resolvedInputSource = this.entityResolver.resolveEntity(publicId, systemId);
 		if (resolvedInputSource == null) {
 			// System.out.println("Resolution result is null.");
 			return null;
 		} else {
-			// System.out.println(MessageFormat.format(
-			// "Resolved to publicId [{0}], systemId [{1}].",
-			// resolvedInputSource.getPublicId(),
-			// resolvedInputSource.getSystemId()));
-			final String pId = publicId != null ? publicId
-					: resolvedInputSource.getPublicId();
-			final String sId = systemId != null ? systemId
-					: resolvedInputSource.getSystemId();
-			return new ReResolvingInputSourceWrapper(this.entityResolver,
-					resolvedInputSource, pId, sId);
+			// System.out.println(MessageFormat.format("Resolved to publicId [{0}], systemId [{1}].", resolvedInputSource.getPublicId(), resolvedInputSource.getSystemId()));
+			final String pId = !StringUtils.isEmpty(publicId) ? publicId : resolvedInputSource.getPublicId();
+			final String sId = computeSystemId(systemId, resolvedInputSource.getSystemId());
+			return new ReResolvingInputSourceWrapper(this.entityResolver, resolvedInputSource, pId, sId, resolvedInputSource.getPublicId(), resolvedInputSource.getSystemId());
 		}
 	}
+
+	private static String computeSystemId(String systemId, String resolvedSystemId) {
+		if (systemId == null) {
+			return resolvedSystemId;
+		}
+		if (resolvedSystemId == null) {
+			return systemId;
+		}
+		boolean fileExistsSystemId = checkFileExists(systemId);
+		boolean fileExistsResolvedSystemId = checkFileExists(resolvedSystemId);
+		return !StringUtils.isEmpty(systemId) && fileExistsSystemId ? systemId : fileExistsResolvedSystemId ? resolvedSystemId : systemId;
+	}
+
+	private static boolean checkFileExists(String sId) {
+		try {
+			URI uriSystemId = new URI(sId);
+			if ("file".equals(uriSystemId.getScheme())) {
+				if (!Files.exists(Paths.get(uriSystemId))) {
+					// resolved file does not exist, warn and let's continue with original systemId
```

Comment:
Should we log a WARN here if user indicates a file:// but the file does not exist? Seems like a swallowed error condition

Reply:
Done too